hat/app/org/hatdex/hat/api/controllers/ContractData.scala

```diff
@@ -0,0 +1,591 @@
+    UserAwareAction.async(parsers.json[ContractDataReadRequest]) { implicit request =>
+      val contractDataRead      = request.body
+      val maybeContractDataInfo = refineContractDataReadRequest(contractDataRead)
+      maybeContractDataInfo match {
+        case Some(contractDataInfo) =>
+          contractValid(contractDataInfo, namespace).flatMap { contractOk =>
+            contractOk match {
+              case (Some(_hatUser @ _), Right(RequestVerified(ns))) =>
+                makeData(ns, endpoint, orderBy, ordering, skip, take)
+              case (_, Left(x)) => handleFailedRequestAssessment(x)
+              case (_, _) =>
+                logger.warn(s"ReadContract: Neither hatUser found or contract verified for ${contractDataRead}")
+                handleFailedRequestAssessment(GeneralError)
+            }
+          }
+        case None => Future.successful(BadRequest("Missing Contract Details."))
+      }
+    }
+
+  def createContractData(
+      namespace: String,
+      endpoint: String,
+      skipErrors: Option[Boolean]): Action[ContractDataCreateRequest] =
+    UserAwareAction.async(parsers.json[ContractDataCreateRequest]) { implicit request =>
+      val contractDataCreate    = request.body
+      val maybeContractDataInfo = refineContractDataCreateRequest(contractDataCreate)
+      maybeContractDataInfo match {
+        case Some(contractDataInfo) =>
+          contractValid(contractDataInfo, namespace).flatMap { contractOk =>
+            contractOk match {
+              case (Some(hatUser), Right(RequestVerified(ns))) =>
+                handleCreateContractData(hatUser, contractDataCreate, ns, endpoint, skipErrors)
+              case (_, Left(x)) => handleFailedRequestAssessment(x)
+              case (_, _) =>
+                logger.warn(s"Neither hatUser found or contract verified for ${contractDataCreate}")
+                handleFailedRequestAssessment(GeneralError)
+            }
+          }
+        case None => Future.successful(BadRequest("Missing Contract Details."))
+      }
+
+      Future.successful(BadRequest("Not Implemented."))
+    }
+
+  def updateContractData(namespace: String): Action[ContractDataUpdateRequest] =
+    UserAwareAction.async(parsers.json[ContractDataUpdateRequest]) { implicit request =>
+      val contractDataUpdate    = request.body
+      val maybeContractDataInfo = refineContractDataUpdateRequest(contractDataUpdate)
+      maybeContractDataInfo match {
+        case Some(contractDataInfo) =>
+          contractValid(contractDataInfo, namespace).flatMap { contractOk =>
+            contractOk match {
+              case (Some(hatUser), Right(RequestVerified(ns))) =>
+                handleUpdateContractData(hatUser, contractDataUpdate, ns)
+              case (_, Left(x)) => handleFailedRequestAssessment(x)
+              case (_, _) =>
+                logger.warn(s"Neither hatUser found or contract verified for ${contractDataUpdate}")
```

Comment:
Non Blocking: deceptive message, a hatUser may of been found it is only contracted that is looked at